nav2_behaviors/src/spin.cpp

```diff
@@ -0,0 +1,174 @@
+  getAnglesFromQuaternion(command->quaternion, yaw, pitch, roll);
+
+  if (roll != 0.0 || pitch != 0.0) {
+    RCLCPP_INFO(get_logger(), "Spinning on Y and X not supported, "
+      "will only spin in Z.");
+  }
+
+  // Testing
+  // if (!getRobotYaw(start_yaw_)) {
+  //   return TaskStatus::FAILED;
+  // }
+
+  RCLCPP_INFO(get_logger(), "Currently only supported spinning by a fixed amount");
+
+  start_time_ = std::chrono::system_clock::now();
+
+  return nav2_tasks::TaskStatus::SUCCEEDED;
+}
+
+nav2_tasks::TaskStatus Spin::onCycleUpdate(nav2_tasks::SpinResult & result)
+{
+  // Currently only an open-loop controller is implemented
+  // TODO(orduno) Create a base class for open-loop controlled behaviors
+  //              controlledSpin() has not been fully tested
+  TaskStatus status = timedSpin();
+
+  // For now sending an empty task result
+  nav2_tasks::SpinResult empty_result;
+  result = empty_result;
+
+
+  return status;
+}
+
+nav2_tasks::TaskStatus Spin::timedSpin()
+{
+  // Output control command
+  geometry_msgs::msg::Twist cmd_vel;
+
+  // TODO(orduno) fixed speed
+  cmd_vel.linear.x = 0.0;
+  cmd_vel.linear.y = 0.0;
+  cmd_vel.angular.z = 0.5;
+  robot_->sendVelocity(cmd_vel);
+
+  // TODO(orduno) fixed time
+  auto current_time = std::chrono::system_clock::now();
+  if (current_time - start_time_ >= 4s) {
+    // Stop the robot
+    cmd_vel.angular.z = 0.0;
+    robot_->sendVelocity(cmd_vel);
+
+    return TaskStatus::SUCCEEDED;
+  }
+
+  return TaskStatus::RUNNING;
+}
+
+nav2_tasks::TaskStatus Spin::controlledSpin()
+{
+  // TODO(orduno) Test and tune controller
+  //              check it doesn't abruptly start and stop
+  //              or cause massive wheel slippage when accelerating
+
+  // Get current robot orientation
+  double current_yaw;
+
+  if (!getRobotYaw(current_yaw)) {
+    return TaskStatus::FAILED;
+  }
+
+  double current_angle = current_yaw - start_yaw_;
+
+  double dist_left = M_PI - current_angle;
+
+  // TODO(orduno) forward simulation to check if future position is feasible
+
+  // compute the velocity that will let us stop by the time we reach the goal
+  // v_f^2 == v_i^2 + 2 * a * d
+  // solving for v_i if v_f = 0
+  double vel = sqrt(2 * rotational_acc_lim_ * dist_left);
+
+  // limit velocity
+  vel = std::min(std::max(vel, min_rotational_vel_), max_rotational_vel_);
+
+  geometry_msgs::msg::Twist cmd_vel;
+  cmd_vel.linear.x = 0.0;
+  cmd_vel.linear.y = 0.0;
+  cmd_vel.angular.z = vel;
+
+  robot_->sendVelocity(cmd_vel);
+
+  // check if we are done
+  if (dist_left >= (0.0 - goal_tolerance_angle_)) {
+    return TaskStatus::SUCCEEDED;
+  }
+
+  return TaskStatus::RUNNING;
+}
+
+void Spin::getAnglesFromQuaternion(
+  const geometry_msgs::msg::Quaternion & quaternion,
+  double & yaw, double & pitch, double & roll)
+{
+  tf2::Matrix3x3(
+    tf2::Quaternion(
+      quaternion.x, quaternion.y, quaternion.z, quaternion.w)).getEulerYPR(yaw, pitch, roll);
+}
+
+bool Spin::getRobotYaw(double & yaw)
```

Comment:
I don't think this method is necessary. tf2::getYaw does this operation.
`yaw = tf2::getYaw(current_pose.pose.orientation)`

Reply:
Thanks, changed